pylint/plugins/hass_constructor.py

```diff
@@ -0,0 +1,53 @@
+from astroid.node_classes import Const
+from astroid.nodes import ClassDef
+from pylint.checkers import BaseChecker
+from pylint.interfaces import IAstroidChecker
+
+
+class HassConstructorFormatChecker(BaseChecker):
+    """Check that __init__ return type is present."""
+
+    __implements__ = IAstroidChecker
+
+    name = "hass_constructor"
+    priority = -1
+    msgs = {
+        "W0006": (
+            '__init__ should have explicit return type "None"',
+            "hass-constructor-return",
+            "Used when __init__ has all arguments typed "
+            "but doesn't have return type declared",
+        ),
+    }
+    options = ()
+
+    def visit_functiondef(self, node):
+        """Called when a :class:`.astroid.node_classes.FunctionDef` node is visited.
+        See :mod:`astroid` for the description of available nodes.
+        :param node: The node to check.
+        :type node: astroid.node_classes.FunctionDef
+        """
+        if not node.is_method() or node.name != "__init__":
+            return
+
+        # Check that all arguments are annotated.
+        # The first argument is "self".
+        annotations = (
+            node.args.posonlyargs_annotations
+            + node.args.annotations
+            + node.args.kwonlyargs_annotations
+        )[1:]
```

Comment:
I haven't checked this particular case, but annotations are a bit tricky. Are you sure `self` is always the first in `posonlyargs_annotations`? Furthermore there is also `varargannotation` and `kwargannotation` which you haven't yet checked (although they might not apply 🤷🏻‍♂️).

Reply:
When positional only args are present, `self` will be first in `posonlyargs_annotations`. If it's not set, `self` will be the first in regular `annotations` while `posonlyargs_annotations` is empty list. This way in concatenated list it's always the first.

I'll check if `varargannotation` and `kwargannotation` apply here.

Reply:
Thanks for pointing out at this, now it should be complete and robust 🙂

Reply:
Maybe 6 entries that aren't properly annotated in the first place?

Reply:
Yes, they have untyped `**kwargs`, so they were filtered out.